client/tso_client.go

```diff
@@ -275,3 +281,114 @@ func (c *tsoClient) backupClientConn() (*grpc.ClientConn, string) {
 	}
 	return nil, ""
 }
+
+// getMinTS gets a timestamp from PD or the minimal timestamp across all keyspace groups from the TSO microservice.
+func (c *tsoClient) getMinTS(ctx context.Context) (physical, logical int64, err error) {
+	// Immediately refresh the TSO server/pod list
+	addrs := c.svcDiscovery.DiscoverMicroservice(tsoService)
+	if len(addrs) == 0 {
+		return 0, 0, errs.ErrClientGetMinTSO.FastGenByArgs("no tso servers/pods discovered")
+	}
+
+	// Get the minimal timestamp from the TSO servers/pods
+	var mutex sync.Mutex
+	resps := make([]*tsopb.GetMinTSResponse, 0)
+	wg := sync.WaitGroup{}
+	wg.Add(len(addrs))
+	for _, addr := range addrs {
+		go func(addr string) {
+			defer wg.Done()
+			resp, err := c.getMinTSFromSingleServer(ctx, addr, c.option.timeout)
```

Comment:
if there is no primary in a tso server, do we need to wait to result for it?

Reply:
Not really, but it does need to wait all keyspace groups in serving status. It needs to meet the following conditions.
	// Check the results. The returned minimal timestamp is valid if all the conditions are met:
	// 1. The number of responses is equal to the number of TSO servers/pods.
	// 2. The number of keyspace groups asked is equal to the number of TSO servers/pods.
	// 3. The minimal timestamp is not zero.

Reply:
We have discussed two more options for further optimization which might be overkilled for now. The two more options we discussed are:
1. Read all saved timestamps of all keyspace groups from etcd. Select the minimal ts then substract a timewindow. The cons is that it could cause stuck in availability when loading all saved timestamps in a transaction. If QPS is higher than frequency of lead start/restart, it could impact availability.
2. Based on the current approach, but every keyspace group cache the last ts. If the update time of the last ts, for a keyspace group, is out of a time window, it needs to query the timestamp to pull forward the timeline. It might be overkill for now.